File: python_scraper/rate_limiter.py
"""
Rate limiting for polite scraping.
"""

import time
import random
import logging
from typing import Callable, Any, Optional


class RateLimiter:
    """Rate limiter with exponential backoff support."""

    def __init__(
        self,
        min_delay: int = 2000,
        max_delay: int = 5000,
        logger: Optional[logging.Logger] = None
    ):
        """
        Initialize rate limiter.

        Args:
            min_delay: Minimum delay in milliseconds
            max_delay: Maximum delay in milliseconds
            logger: Logger instance
        """
        self.min_delay = min_delay / 1000.0  # Convert to seconds
        self.max_delay = max_delay / 1000.0
        self.last_request_time: Optional[float] = None
        self.logger = logger or logging.getLogger(__name__)

    def wait(self) -> None:
        """Wait with random delay between min and max."""
        delay = random.uniform(self.min_delay, self.max_delay)

        # Add jitter for human-like behavior
        jitter = random.uniform(-0.2, 0.2) * delay
        total_delay = max(0, delay + jitter)

        if self.last_request_time:
            # Ensure minimum delay since last request
            elapsed = time.time() - self.last_request_time
            if elapsed < total_delay:
                time.sleep(total_delay - elapsed)
        else:
            time.sleep(total_delay)

        self.last_request_time = time.time()

    def retry_with_backoff(
        self,
        func: Callable,
        max_retries: int = 3,
        *args,
        **kwargs
    ) -> Any:
        """
        Retry function with exponential backoff.

        Args:
            func: Function to retry
            max_retries: Maximum number of retries
            *args: Positional arguments for func
            **kwargs: Keyword arguments for func

        Returns:
            Function result

        Raises:
            Last exception if all retries fail
        """
        last_exception = None

        for attempt in range(max_retries):
            try:
                return func(*args, **kwargs)
            except Exception as e:
                last_exception = e
                if attempt < max_retries - 1:
                    # Exponential backoff: 2^attempt seconds
                    backoff_time = 2 ** attempt
                    self.logger.warning(
                        f"Attempt {attempt + 1} failed: {e}. "
                        f"Retrying in {backoff_time}s..."
                    )
                    time.sleep(backoff_time)
                else:
                    self.logger.error(f"All {max_retries} attempts failed")

        raise last_exception

File: src/core/rate-limiter.js
// FIXED: Constructor now accepts logger and options to match usage in orchestrator.js
class RateLimiter {
  constructor(logger = null, options = {}) {
    this.logger = logger;
    this.minDelay = options.minDelay || 2000; // 2 seconds minimum
    this.maxDelay = options.maxDelay || 5000; // 5 seconds maximum
    this.backoffMultiplier = 1.5;
    this.maxRetries = 3;
    this.lastRequestTime = 0;
  }

  /**
   * Wait before making the next request
   * Adds random delay between minDelay and maxDelay for human-like behavior
   */
  async waitBeforeRequest() {
    const now = Date.now();
    const timeSinceLastRequest = now - this.lastRequestTime;

    // Calculate random delay with jitter
    const randomDelay = Math.random() * (this.maxDelay - this.minDelay) + this.minDelay;

    // If not enough time has passed, wait for the remaining time plus random jitter
    if (timeSinceLastRequest < randomDelay) {
      const waitTime = randomDelay - timeSinceLastRequest;
      // FIXED: Use this.logger instead of imported logger
      if (this.logger) {
        this.logger.debug(`Rate limiting: waiting ${(waitTime / 1000).toFixed(2)}s before next request`);
      }
      await this.sleep(waitTime);
    }

    this.lastRequestTime = Date.now();
  }

  /**
   * Retry a function with exponential backoff
   * @param {Function} fn - Async function to retry
   * @param {string} context - Description of the operation for logging
   * @returns {Promise} Result of the function
   */
  async retryWithBackoff(fn, context = 'operation') {
    let lastError;
    let currentDelay = this.minDelay;

    for (let attempt = 1; attempt <= this.maxRetries; attempt++) {
      try {
        // FIXED: Use this.logger with null check
        if (this.logger) {
          this.logger.debug(`${context} - Attempt ${attempt}/${this.maxRetries}`);
        }

        // Wait before request (except first attempt)
        if (attempt > 1) {
          const backoffDelay = currentDelay * Math.pow(this.backoffMultiplier, attempt - 1);
          const jitteredDelay = backoffDelay * (0.8 + Math.random() * 0.4); // +/- 20% jitter

          if (this.logger) {
            this.logger.info(`Retrying ${context} after ${(jitteredDelay / 1000).toFixed(2)}s (attempt ${attempt}/${this.maxRetries})`);
          }
          await this.sleep(jitteredDelay);
        }

        // Execute the function
        const result = await fn();

        // Success - reset delay and return
        if (attempt > 1 && this.logger) {
          this.logger.info(`${context} succeeded on attempt ${attempt}`);
        }
        return result;

      } catch (error) {
        lastError = error;
        if (this.logger) {
          this.logger.warn(`${context} failed (attempt ${attempt}/${this.maxRetries}): ${error.message}`);
        }

        // Don't retry CAPTCHA errors
        if (error.message === 'CAPTCHA_DETECTED') {
          throw error;
        }

        // If this was the last attempt, throw the error
        if (attempt === this.maxRetries) {
          if (this.logger) {
            this.logger.error(`${context} failed after ${this.maxRetries} attempts`);
          }
          throw lastError;
        }
      }
    }

    // Should never reach here, but just in case
    throw lastError;
  }

  /**
   * Sleep for specified milliseconds
   * @param {number} ms - Milliseconds to sleep
   */
  sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  /**
   * Add random human-like delay
   * Useful for simulating user actions like scrolling or clicking
   * @param {number} baseMs - Base delay in milliseconds
   * @param {number} variance - Variance percentage (0-1)
   */
  async randomDelay(baseMs = 1000, variance = 0.3) {
    const minMs = baseMs * (1 - variance);
    const maxMs = baseMs * (1 + variance);
    const delay = Math.random() * (maxMs - minMs) + minMs;

    // FIXED: Use this.logger with null check
    if (this.logger) {
      this.logger.debug(`Random delay: ${(delay / 1000).toFixed(2)}s`);
    }
    await this.sleep(delay);
  }

  /**
   * Set custom delay range
   * @param {number} minMs - Minimum delay in milliseconds
   * @param {number} maxMs - Maximum delay in milliseconds
   */
  setDelayRange(minMs, maxMs) {
    if (minMs < 0 || maxMs < minMs) {
      throw new Error('Invalid delay range');
    }
    this.minDelay = minMs;
    this.maxDelay = maxMs;
    // FIXED: Use this.logger with null check
    if (this.logger) {
      this.logger.info(`Rate limiter delay range updated: ${minMs}-${maxMs}ms`);
    }
  }

  /**
   * Set maximum number of retries
   * @param {number} retries - Number of retries (1-10)
   */
  setMaxRetries(retries) {
    if (retries < 1 || retries > 10) {
      throw new Error('Max retries must be between 1 and 10');
    }
    this.maxRetries = retries;
    // FIXED: Use this.logger with null check
    if (this.logger) {
      this.logger.info(`Max retries updated: ${retries}`);
    }
  }

  /**
   * Reset the last request time
   * Useful when starting a new scraping session
   */
  reset() {
    this.lastRequestTime = 0;
    // FIXED: Use this.logger with null check
    if (this.logger) {
      this.logger.debug('Rate limiter reset');
    }
  }

  /**
   * Alias for waitBeforeRequest() - for backward compatibility
   * Some scrapers use rateLimiter.wait() instead of waitBeforeRequest()
   */
  async wait() {
    return this.waitBeforeRequest();
  }
}

module.exports = RateLimiter;
